File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/UpdatePasswordFragment.java
package org.mifos.mobilebanking.ui.fragments;

/*
 * Created by saksham on 13/July/2018
 */

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.models.UpdatePasswordPayload;
import org.mifos.mobilebanking.presenters.UpdatePasswordPresenter;
import org.mifos.mobilebanking.ui.activities.SettingsActivity;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.UpdatePasswordView;
import org.mifos.mobilebanking.utils.Toaster;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class UpdatePasswordFragment extends BaseFragment implements UpdatePasswordView,
        TextWatcher, View.OnFocusChangeListener {

    @BindView(R.id.til_new_password)
    TextInputLayout tilNewPassword;

    @BindView(R.id.til_confirm_new_password)
    TextInputLayout tilConfirmNewPassword;

    @Inject
    UpdatePasswordPresenter presenter;

    @Inject
    PreferencesHelper preferencesHelper;

    private View rootView;
    private UpdatePasswordPayload payload;
    private boolean isFocusLostNewPassword = false;
    private boolean isFocusLostConfirmPassword = false;

    public static UpdatePasswordFragment newInstance() {
        return new UpdatePasswordFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_update_password, container, false);
        setToolbarTitle(getString(R.string.change_password));
        ButterKnife.bind(this, rootView);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        presenter.attachView(this);

        tilNewPassword.getEditText().addTextChangedListener(this);
        tilConfirmNewPassword.getEditText().addTextChangedListener(this);
        tilNewPassword.getEditText().setOnFocusChangeListener(this);
        tilConfirmNewPassword.getEditText().setOnFocusChangeListener(this);
        return rootView;
    }

    @OnClick(R.id.btn_update_password)
    void updatePassword() {
        if (isFieldsCompleted()) {
            presenter.updateAccountPassword(getUpdatePasswordPayload());
        }
    }

    private boolean isFieldsCompleted() {
        boolean rv = true;
        String newPassword = tilNewPassword.getEditText().getText().toString().trim();
        String repeatPassword = tilConfirmNewPassword.getEditText().getText().toString().trim();

        if (!checkNewPasswordFieldsComplete()) {
            rv = false;
        }
        if (!checkConfirmPasswordFieldsComplete()) {
            rv = false;
        }
        if (!newPassword.equals(repeatPassword)) {
            Toaster.show(rootView, getString(R.string.error_password_not_match));
            rv = false;
        }
        return rv;
    }

    private UpdatePasswordPayload getUpdatePasswordPayload() {
        payload = new UpdatePasswordPayload();
        payload.setPassword(tilNewPassword.getEditText().getText().toString().trim());
        payload.setRepeatPassword(tilConfirmNewPassword.getEditText().getText().toString().trim());
        return payload;
    }

    @Override
    public void showError(String message) {
        Toaster.show(rootView, message);
    }

    @Override
    public void showPasswordUpdatedSuccessfully() {
        Toast.makeText(getContext(), getString(R.string.string_changed_successfully,
                getString(R.string.password)), Toast.LENGTH_SHORT).show();
        startActivity(new Intent(getContext(), SettingsActivity.class));
    }

    @Override
    public void showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading));
    }

    @Override
    public void hideProgress() {
        hideMifosProgressDialog();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.detachView();
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        if (tilNewPassword.getEditText().hasFocus() && isFocusLostNewPassword) {
            checkNewPasswordFieldsComplete();
        }
        if (tilConfirmNewPassword.getEditText().hasFocus() && isFocusLostConfirmPassword) {
            checkConfirmPasswordFieldsComplete();
        }
    }

    @Override
    public void afterTextChanged(Editable s) {

    }

    @Override
    public void onFocusChange(View v, boolean hasFocus) {
        if (v.getId() == R.id.et_new_password && !isFocusLostNewPassword && !hasFocus) {
            checkNewPasswordFieldsComplete();
            isFocusLostNewPassword = true;
        }
        if (v.getId() == R.id.et_confirm_password && !isFocusLostConfirmPassword && !hasFocus) {
            checkConfirmPasswordFieldsComplete();
            isFocusLostConfirmPassword = true;
        }
    }

    private boolean checkNewPasswordFieldsComplete() {
        String newPassword = tilNewPassword.getEditText().getText().toString();
        isFocusLostNewPassword = true;
        if (newPassword.isEmpty()) {
            tilNewPassword.setError(
                    getString(R.string.error_validation_blank,
                            getString(R.string.new_password)));
            return false;
        }
        if (newPassword.length() < 6) {
            tilNewPassword.setError(
                    getString(R.string.error_validation_minimum_chars,
                            getString(R.string.new_password),
                            getResources().getInteger(R.integer.password_minimum_length)));
            return false;
        }
        tilNewPassword.setErrorEnabled(false);
        return true;
    }

    private boolean checkConfirmPasswordFieldsComplete() {
        String confirmPassword = tilConfirmNewPassword.getEditText().getText().toString();
        isFocusLostConfirmPassword = true;
        if (confirmPassword.isEmpty()) {
            tilConfirmNewPassword.setError(
                    getString(R.string.error_validation_blank,
                            getString(R.string.confirm_password)));
            return false;
        }
        if (confirmPassword.length() < 6) {
            tilConfirmNewPassword.setError(
                    getString(R.string.error_validation_minimum_chars,
                            getString(R.string.confirm_password),
                            getResources().getInteger(R.integer.password_minimum_length)));
            return false;
        }
        tilConfirmNewPassword.setErrorEnabled(false);
        return true;
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/UpdatePasswordFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentUpdatePasswordBinding
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.MFErrorParser
import org.mifos.mobile.utils.Network
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.utils.RegistrationUiState
import org.mifos.mobile.viewModels.UpdatePasswordViewModel

/*
* Created by saksham on 13/July/2018
*/
@AndroidEntryPoint
class UpdatePasswordFragment : BaseFragment(), TextWatcher, OnFocusChangeListener {

    private var _binding: FragmentUpdatePasswordBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: UpdatePasswordViewModel
    private var isFocusLostNewPassword = false
    private var isFocusLostConfirmPassword = false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentUpdatePasswordBinding.inflate(inflater, container, false)
        setToolbarTitle(getString(R.string.change_password))
        viewModel = ViewModelProvider(this)[UpdatePasswordViewModel::class.java]
        binding.tilNewPassword.editText?.addTextChangedListener(this)
        binding.tilConfirmNewPassword.editText?.addTextChangedListener(this)
        binding.tilNewPassword.editText?.onFocusChangeListener = this
        binding.tilConfirmNewPassword.editText?.onFocusChangeListener = this
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnUpdatePassword.setOnClickListener {
            updatePassword()
        }

        viewModel.updatePasswordUiState.observe(viewLifecycleOwner) { state ->
            when (state) {
                RegistrationUiState.Loading -> showProgress()
                RegistrationUiState.Success -> {
                    hideProgress()
                    showPasswordUpdatedSuccessfully()
                }

                is RegistrationUiState.Error -> {
                    hideProgress()
                    showError(MFErrorParser.errorMessage(state.exception))
                }
            }
        }
    }

    private fun updatePassword() {
        val newPassword = binding.tilNewPassword.editText?.text.toString().trim()
        val confirmPassword = binding.tilConfirmNewPassword.editText?.text.toString().trim()

        if (areFieldsValidated(newPassword, confirmPassword)) {
            viewModel.updateAccountPassword(newPassword, confirmPassword)
        }
    }

    private fun areFieldsValidated(newPassword: String, confirmPassword: String): Boolean {
        return when {
            !isNewPasswordValidated() -> false
            !isConfirmPasswordValidated() -> false
            (!viewModel.validatePasswordMatch(newPassword, confirmPassword)) -> {
                Toaster.show(binding.root, getString(R.string.error_password_not_match))
                false
            }

            else -> true
        }
    }

    private fun isNewPasswordValidated(): Boolean {
        with(binding) {
            val newPassword = tilNewPassword.editText?.text.toString()
            isFocusLostNewPassword = true
            return when {
                viewModel.isInputFieldEmpty(newPassword) -> {
                    tilNewPassword.error = getString(
                        R.string.error_validation_blank,
                        getString(R.string.new_password),
                    )
                    false
                }

                viewModel.isInputLengthInadequate(newPassword) -> {
                    tilNewPassword.error = getString(
                        R.string.error_validation_minimum_chars,
                        getString(R.string.new_password),
                        resources.getInteger(R.integer.password_minimum_length),
                    )
                    false
                }

                else -> {
                    tilNewPassword.isErrorEnabled = false
                    return true
                }
            }
        }
    }

    private fun isConfirmPasswordValidated(): Boolean {
        with(binding) {
            val confirmPassword = tilConfirmNewPassword.editText?.text.toString()
            isFocusLostConfirmPassword = true
            return when {
                viewModel.isInputFieldEmpty(confirmPassword) -> {
                    tilConfirmNewPassword.error = getString(
                        R.string.error_validation_blank,
                        getString(R.string.confirm_password),
                    )
                    false
                }

                viewModel.isInputLengthInadequate(confirmPassword) -> {
                    tilConfirmNewPassword.error = getString(
                        R.string.error_validation_minimum_chars,
                        getString(R.string.confirm_password),
                        resources.getInteger(R.integer.password_minimum_length),
                    )
                    return false
                }

                else -> {
                    tilConfirmNewPassword.isErrorEnabled = false
                    return true
                }
            }
        }

    }

    fun showError(message: String?) {
        var errorMessage = message
        if (!Network.isConnected(activity)) {
            errorMessage = getString(R.string.no_internet_connection)
        }
        Toaster.show(binding.root, errorMessage)
    }

    private fun showPasswordUpdatedSuccessfully() {
        Toast.makeText(
            context,
            getString(
                R.string.string_changed_successfully,
                getString(R.string.password),
            ),
            Toast.LENGTH_SHORT,
        ).show()
        (activity as BaseActivity).clearFragmentBackStack()
        (activity as BaseActivity).replaceFragment(
            SettingsFragment.newInstance(),
            true,
            R.id.container,
        )
    }

    fun showProgress() {
        showMifosProgressDialog(getString(R.string.progress_message_loading))
    }

    fun hideProgress() {
        hideMifosProgressDialog()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        if (binding.tilNewPassword.editText?.hasFocus() == true && isFocusLostNewPassword) {
            isNewPasswordValidated()
        }
        if (binding.tilConfirmNewPassword.editText?.hasFocus() == true && isFocusLostConfirmPassword) {
            isConfirmPasswordValidated()
        }
    }

    override fun afterTextChanged(s: Editable) {}
    override fun onFocusChange(v: View, hasFocus: Boolean) {
        if (v.id == R.id.et_new_password && !isFocusLostNewPassword && !hasFocus) {
            isNewPasswordValidated()
            isFocusLostNewPassword = true
        }
        if (v.id == R.id.et_confirm_password && !isFocusLostConfirmPassword && !hasFocus) {
            isConfirmPasswordValidated()
            isFocusLostConfirmPassword = true
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(): UpdatePasswordFragment {
            return UpdatePasswordFragment()
        }
    }
}
